app/models/specific_tube_creation.rb

```diff
@@ -64,13 +64,44 @@ def connect_parent_and_children
   end
   private :connect_parent_and_children
 
+  def determine_foreign_barcode_format(foreign_barcode)
+    Barcode.matching_barcode_format(foreign_barcode)
+  end
+
+  def check_foreign_barcode_unique(foreign_barcode_format, foreign_barcode)
+    return unless Barcode.exists_for_format?(foreign_barcode_format, foreign_barcode)
+
+    raise "Foreign Barcode: #{foreign_barcode} is already in use!"
+  end
+
+  def add_foreign_barcode_to_tube(tube, foreign_barcode)
+    # determine the format of the foreign barcode
+    foreign_barcode_format = determine_foreign_barcode_format(foreign_barcode)
+
+    raise "Cannot determine format for foreign barcode #{foreign_barcode}" if foreign_barcode_format.blank?
+
+    # check for uniqueness
```

Comment:
I think you could remove most of the comments you've added here - the methods are v well named so they are just duplicating information.